packages/angular/cli/models/schematic-engine-host.ts

```diff
@@ -0,0 +1,122 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { RuleFactory } from '@angular-devkit/schematics';
+import { NodeModulesEngineHost } from '@angular-devkit/schematics/tools';
+import { readFileSync } from 'fs';
+import { dirname } from 'path';
+import { Script } from 'vm';
+
+export class SchematicEngineHost extends NodeModulesEngineHost {
+  protected _resolveReferenceString(refString: string, parentPath: string) {
+    const [path, name] = refString.split('#', 2);
+    const schematicFile = require.resolve(path, { paths: [parentPath] });
+
+    const isAngularSchematic = /[\/\\]node_modules[\/\\]@(?:angular|schematics)[\/\\]/.test(
+      schematicFile,
+    );
+
+    // Angular schematics are safe to use in the wrapped VM context
+    if (isAngularSchematic) {
+      const schematicPath = dirname(schematicFile);
+
+      const moduleCache = new Map<string, unknown>();
+      const factoryInitializer = wrap(
+        schematicFile,
+        schematicPath,
+        moduleCache,
+        name || 'default',
+      ) as () => RuleFactory<{}>;
+
+      const factory = factoryInitializer();
+      if (!factory || typeof factory !== 'function') {
+        return null;
+      }
+
+      return { ref: factory, path: schematicPath };
+    }
+
+    // All other schematics use default behavior
+    return super._resolveReferenceString(refString, parentPath);
+  }
+}
+
+/**
+ * Wrap a JavaScript file in a VM context to allow specific Angular dependencies to be redirected.
+ * This VM setup is ONLY intended to redirect dependencies.
+ *
+ * @param schematicFile A JavaScript schematic file path that should be wrapped.
+ * @param schematicDirectory A directory that will be used as the location of the JavaScript file.
+ * @param moduleCache A map to use for caching repeat module usage and proper `instanceof` support.
+ * @param exportName An optional name of a specific export to return. Otherwise, return all exports.
+ */
+function wrap(
+  schematicFile: string,
+  schematicDirectory: string,
+  moduleCache: Map<string, unknown>,
+  exportName?: string,
+): () => unknown {
+  const { createRequire, createRequireFromPath } = require('module');
+  // Node.js 10.x does not support `createRequire` so fallback to `createRequireFromPath`
+  // `createRequireFromPath` is deprecated in 12+ and can be removed once 10.x support is removed
+  const scopedRequire = createRequire?.(schematicFile) || createRequireFromPath(schematicDirectory);
+
+  const customRequire = function (id: string) {
+    if (id.startsWith('@angular-devkit/') || id.startsWith('@schematics/')) {
```

Comment:
I think this logic might cause problems when using Angular packages which rely on older versions of Angular schematics packages.

Something like;
```
ng update @angular/core@8
```

Previously, users could have worked around this problem by installing the correct version of the dependency in their project, Example ` npm i @angular-devkit/schematics@8`, but now this will break, since the dependency will always be resolved from the CLI package would always be the latest.

Reply:
`@angular/core` is actually functional in this regard.  `@angular/material` is, however, broken currently in those situations (with or without this PR).  The direct package install is a potential workaround for those cases but it can also cause other breakage since the object instances given to the schematic will be from the latest versions but the functions acting on them will be from an older version which might assume different internal behavior/structure so it's not an ideal solution.
An environment variable will be available (`NG_SCHEMATIC_REDIRECT=0`) which will allow the workaround to continue for those edge cases with old updates using new CLI versions.  Updates before version 8 do not have the update bootstrap feature so the changes in this PR will not apply.
This PR also provides the base capabilities to potentially remove the need for any workarounds by providing virtual compatibility modules for the deprecated and removed imports.

Reply:
Minimal shim modules for the legacy deep imports into `@schematics/angular` are now available for the old first-party schematics that previously used them.  This should remove the need for the manual package install workaround in the latest CLI updating to old package versions case.